base_layer/core/src/base_node/comms_interface/inbound_handlers.rs

```diff
@@ -113,16 +123,34 @@ where T: BlockchainBackend
                 )
                 .await?,
             )),
-            NodeCommsRequest::GetNewBlock =>
-            // TODO: query blockchain_db and mempool to construct a new mineable block
-            {
-                unimplemented!()
+            NodeCommsRequest::GetNewBlock => {
+                let metadata = async_db::get_metadata(self.blockchain_db.clone()).await?;
+                let best_block_hash = metadata.best_block.ok_or(CommsInterfaceError::UnexpectedApiResponse)?;
+                let best_block_header =
+                    async_db::fetch_header_with_block_hash(self.blockchain_db.clone(), best_block_hash).await?;
+                let mut header = BlockHeader::from_previous(&best_block_header);
+                header.total_difficulty = best_block_header.total_difficulty + Difficulty::from(1); // TODO: the new blocks difficulty should be calculated using lwma_diff
+
+                let transactions = self
+                    .mempool
+                    .retrieve(MAX_BLOCK_TRANSACTION_WEIGHT)
+                    .map_err(|e| CommsInterfaceError::MempoolError(e.to_string()))?
+                    .iter()
+                    .map(|tx| (**tx).clone())
+                    .collect();
+
+                let block = BlockBuilder::new()
+                    .with_header(header)
+                    .with_transactions(transactions)
+                    .build();
+                Ok(NodeCommsResponse::NewBlock(block))
             },
         }
     }
 
     /// Handle inbound blocks from remote nodes and local services.
     pub async fn handle_block(&mut self, block: &Block) -> Result<(), CommsInterfaceError> {
+        // TODO perform additional validation of the block before calling add_block
```

Comment:
The validation pipeline will handle this

Reply:
Perfect, I removed the comment.